File: java-batch/src/main/java/com/neoxam/batch/SynthesisService.java
package com.neoxam.batch;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Génère un rapport de synthèse en texte brut à partir d'une liste de
 * lignes { label, amount }. Portage Java de la fonctionnalité décrite
 * dans specs/rapport-synthese.md (fonction pure côté TypeScript).
 *
 * Le total est calculé via BatchReportService.aggregateTotal() : ce
 * service hérite donc, par choix explicite, du bug #5 connu (dérive
 * flottante sur de grands volumes de lignes) plutôt que de recalculer
 * la somme indépendamment.
 */
public class SynthesisService {

    private static final String TITRE_PAR_DEFAUT = "SYNTHÈSE";
    private static final String SEPARATEUR = "---";

    private final BatchReportService batchReportService = new BatchReportService();

    public String generateSynthesis(List<SynthesisLine> lines) {
        return generateSynthesis(lines, null);
    }

    public String generateSynthesis(List<SynthesisLine> lines, SynthesisOptions options) {
        String title = (options != null && options.getTitle() != null)
                ? options.getTitle()
                : TITRE_PAR_DEFAUT;

        StringBuilder sb = new StringBuilder();
        sb.append("=== ").append(title).append(" ===\n");
        for (SynthesisLine line : lines) {
            sb.append(String.format("%s: %.2f %s", line.getLabel(), line.getAmount(), "€")).append('\n');
        }
        sb.append(SEPARATEUR).append('\n');

        int count = lines.size();
        double total = batchReportService.aggregateTotal(toReportLines(lines));
        double average = count == 0 ? 0.0 : total / count;

        sb.append("Nombre de lignes: ").append(count).append('\n');
        sb.append(String.format("Total: %.2f %s", total, "€")).append('\n');
        sb.append(String.format("Moyenne: %.2f %s", average, "€"));

        return sb.toString();
    }

    private List<ReportLine> toReportLines(List<SynthesisLine> lines) {
        return lines.stream()
                .map(line -> new ReportLine(line.getLabel(), line.getAmount()))
                .collect(Collectors.toList());
    }
}

File: src/services/synthesis.service.ts
/**
 * Service de génération de rapport de synthèse en texte brut.
 *
 * Fonction pure et réutilisable : à partir d'une liste de lignes
 * { label, amount }, produit un texte récapitulatif (détail des lignes,
 * total, moyenne). Ne dépend d'aucun framework HTTP ni du module
 * LegacyRenderer (en attente de refactoring).
 *
 * Cohérence du total : le rapport doit « boucler », c'est-à-dire que le total
 * affiché doit être exactement égal à la somme des lignes affichées. Cela
 * impose de n'arrondir qu'UNE seule fois, au centime, au niveau de la ligne,
 * puis de cumuler ces entiers. Cumuler les montants bruts avant d'arrondir
 * (`Σ montant` puis `.toFixed(2)`) fait apparaître un écart de quelques
 * centimes dès que les montants d'entrée ont une précision sous-centime :
 * chaque ligne affichée est décalée d'un demi-centime au plus, et ces écarts
 * s'accumulent sur un grand nombre de lignes.
 */

export interface SynthesisLine {
  label: string;
  amount: number;
}

export interface SynthesisOptions {
  title?: string;
}

const DEFAULT_TITLE = 'SYNTHÈSE';

/**
 * Convertit un montant en euros en un nombre entier de centimes.
 *
 * Unique point d'arrondi du service : lignes, total et moyenne sont tous
 * dérivés de cette valeur. Même convention que `calculateTTC`
 * (`invoice.service.ts`) pour rester homogène dans le dépôt.
 */
function versCentimes(montant: number): number {
  return Math.round(montant * 100);
}

/**
 * Formate un nombre entier de centimes en euros. Le calcul se fait sur des
 * entiers, sans repasser par un flottant, donc sans nouvel arrondi possible.
 */
function formaterCentimes(centimes: number): string {
  // Entrées dégénérées (NaN, Infinity) : on conserve le rendu d'origine.
  if (!Number.isFinite(centimes)) return (centimes / 100).toFixed(2);

  const signe = centimes < 0 ? '-' : '';
  const absolu = Math.abs(centimes);
  return `${signe}${Math.floor(absolu / 100)}.${String(absolu % 100).padStart(2, '0')}`;
}

export function generateSynthesis(lines: SynthesisLine[], options?: SynthesisOptions): string {
  const title = options?.title ?? DEFAULT_TITLE;

  // Arrondi au centime une fois par ligne, puis cumul sur des entiers :
  // ni écart entre le total et les lignes, ni dérive d'accumulation IEEE 754.
  const centimesParLigne = lines.map((line) => versCentimes(line.amount));
  const totalCentimes = centimesParLigne.reduce((sum, centimes) => sum + centimes, 0);
  const moyenneCentimes = lines.length === 0 ? 0 : Math.round(totalCentimes / lines.length);

  const output: string[] = [`=== ${title} ===`];

  lines.forEach((line, index) => {
    output.push(`${line.label}: ${formaterCentimes(centimesParLigne[index])} €`);
  });

  output.push('---');
  output.push(`Nombre de lignes: ${lines.length}`);
  output.push(`Total: ${formaterCentimes(totalCentimes)} €`);
  // La moyenne est un indicateur dérivé, pas un terme de la somme : elle est
  // arrondie au centime et ne vérifie donc pas `moyenne × n = total`
  // (1.00 / 3 → 0.33, et 3 × 0.33 = 0.99). C'est arithmétiquement inévitable.
  output.push(`Moyenne: ${formaterCentimes(moyenneCentimes)} €`);

  return output.join('\n');
}
